pkg/resourcekeeper/gc.go

```diff
@@ -273,3 +286,68 @@ func (h *gcHandler) GarbageCollectComponentRevisionResourceTracker(ctx context.C
 	}
 	return nil
 }
+
+func (h *gcHandler) GarbageCollectLegacyResourceTrackers(ctx context.Context) error {
+	// skip legacy gc if application is not handled by new version rt
+	if h.app.GetDeletionTimestamp() == nil && h.resourceKeeper._currentRT == nil {
+		return nil
+	}
+	// check app version
+	velaVersionNumberToUpgradeResourceTracker := "v1.2.0"
+	velaVersionToUpgradeResourceTracker, _ := version.NewVersion(velaVersionNumberToUpgradeResourceTracker)
+	var currentVersionNumber string
+	if annotations := h.app.GetAnnotations(); annotations != nil && annotations[oam.AnnotationKubeVelaVersion] != "" {
+		currentVersionNumber = annotations[oam.AnnotationKubeVelaVersion]
+	}
+	currentVersion, err := version.NewVersion(currentVersionNumber)
+	if err == nil && velaVersionToUpgradeResourceTracker.LessThanOrEqual(currentVersion) {
+		return nil
+	}
+	// remove legacy ResourceTrackers
+	clusters := map[string]bool{multicluster.ClusterLocalName: true}
+	for _, rsc := range h.app.Status.AppliedResources {
+		if rsc.Cluster != "" {
+			clusters[rsc.Cluster] = true
+		}
+	}
+	for _, policy := range h.app.Spec.Policies {
+		if policy.Type == v1alpha1.EnvBindingPolicyType {
+			spec := &v1alpha1.EnvBindingSpec{}
+			if err = json.Unmarshal(policy.Properties.Raw, &spec); err == nil {
+				for _, env := range spec.Envs {
+					if env.Placement.ClusterSelector != nil && env.Placement.ClusterSelector.Name != "" {
+						clusters[env.Placement.ClusterSelector.Name] = true
+					}
+				}
+			}
+		}
+	}
+	for cluster := range clusters {
+		_ctx := multicluster.ContextWithClusterName(ctx, cluster)
+		rts := &unstructured.UnstructuredList{}
+		rts.SetGroupVersionKind(v1beta1.SchemeGroupVersion.WithKind("ResourceTracker"))
+		if err = h.Client.List(_ctx, rts, client.MatchingLabels(map[string]string{
+			oam.LabelAppName:      h.app.Name,
+			oam.LabelAppNamespace: h.app.Namespace,
+		})); err != nil {
+			return errors.Wrapf(err, "failed to list resource trackers for app %s/%s in cluster %s", h.app.Namespace, h.app.Name, cluster)
+		}
+		for _, rt := range rts.Items {
+			if s, exists, _ := unstructured.NestedString(rt.Object, "spec", "type"); !exists || s == "" {
+				if err = h.Client.Delete(_ctx, rt.DeepCopy()); err != nil {
+					return errors.Wrapf(err, "failed to delete legacy resource tracker %s for app %s/%s in cluster %s", rt.GetName(), h.app.Namespace, h.app.Name, cluster)
+				}
+			}
+		}
+	}
+	// upgrade app version
+	upgradedVersion := version2.VelaVersion
+	if upgradedVersion == "UNKNOWN" {
+		upgradedVersion = velaVersionNumberToUpgradeResourceTracker
+	}
```

Comment:
you just need to return for the UNKNOWN case, don't add a fixed magic number here.